arch/x86/core/intel64/locore.S

```diff
@@ -0,0 +1,599 @@
+EXCEPT      (24); EXCEPT      (25); EXCEPT      (26); EXCEPT      (27)
+EXCEPT      (28); EXCEPT      (29); EXCEPT      (30); EXCEPT      (31)
+
+/*
+ * When we arrive at 'irq' from one of the IRQ(X) stubs,
+ * we're on the "freshest" IRQ stack and it contains:
+ *
+ *   SS
+ *   RSP
+ *   RFLAGS
+ *   CS
+ *   RIP
+ *   (vector number - IV_IRQS) <-- RSP points here
+ *   RSI <-- we push this on entry
+ */
+
+.globl x86_irq_funcs	/* see irq_manage.c .. */
+.globl x86_irq_args	/* .. for these definitions */
+
+irq:
+	pushq %rsi
+	movq $_kernel, %rsi
+
+	/*
+	 * Bump the IRQ nesting count and move to the next IRQ stack.
+	 * That's sufficient to safely re-enable interrupts, so if we
+	 * haven't reached the maximum nesting depth yet, do it.
+	 */
+
+	incl _kernel_offset_to_nested(%rsi)
+	addq $CONFIG_ISR_STACK_SIZE, ist1
+	cmpl $CONFIG_ISR_DEPTH, _kernel_offset_to_nested(%rsi)
+	jz 1f
+	sti
+1:	cmpl $1, _kernel_offset_to_nested(%rsi)
+	je irq_enter_unnested
+
+	/*
+	 * if we're a nested interrupt, we have to dump the state to the
+	 * stack. we play some games here to re-arrange the stack thusly:
+	 *
+	 * SS RSP RFLAGS CS RIP RAX RSI RCX RDX RBX
+	 * RDI RBP R8 R9 R10 R11 R12 R13 R14 R15
+	 * X86_FXSAVE_SIZE bytes of SSE data <-- RSP points here
+	 *
+	 * note that the final value of RSP must be 16-byte aligned here,
+	 * both to satisfy FXSAVE/FXRSTOR but also to honor the C ABI.
+	 */
+
+irq_enter_nested: /* Nested IRQ: dump register state to stack. */
+	pushq %rcx
+	movq 16(%rsp), %rcx /* RCX = vector */
+	movq %rax, 16(%rsp) /* looks like we pushed RAX, not the vector */
+	pushq %rdx
+	pushq %rbx
+	pushq %rdi
+	pushq %rbp
+	pushq %r8
+	pushq %r9
+	pushq %r10
+	pushq %r11
+	pushq %r12
+	pushq %r13
+	pushq %r14
+	pushq %r15
+	subq $X86_FXSAVE_SIZE, %rsp
+	fxsave (%rsp)
+	jmp irq_dispatch
+
+irq_enter_unnested: /* Not nested: dump state to thread struct for __resume */
+	movq _kernel_offset_to_current(%rsi), %rsi
+	andb $(~_THREAD_SWAPPED), _thread_offset_to_thread_state(%rsi)
+	fxsave _thread_offset_to_sse(%rsi)
+	movq %rbx, _thread_offset_to_rbx(%rsi)
+	movq %rbp, _thread_offset_to_rbp(%rsi)
+	movq %r12, _thread_offset_to_r12(%rsi)
+	movq %r13, _thread_offset_to_r13(%rsi)
+	movq %r14, _thread_offset_to_r14(%rsi)
+	movq %r15, _thread_offset_to_r15(%rsi)
+	movq %rax, _thread_offset_to_rax(%rsi)
+	movq %rcx, _thread_offset_to_rcx(%rsi)
+	movq %rdx, _thread_offset_to_rdx(%rsi)
+	movq %rdi, _thread_offset_to_rdi(%rsi)
+	movq %r8, _thread_offset_to_r8(%rsi)
+	movq %r9, _thread_offset_to_r9(%rsi)
+	movq %r10, _thread_offset_to_r10(%rsi)
+	movq %r11, _thread_offset_to_r11(%rsi)
+	popq %rax /* RSI */
+	movq %rax, _thread_offset_to_rsi(%rsi)
+	popq %rcx /* vector number */
+	popq %rax /* RIP */
+	movq %rax, _thread_offset_to_rip(%rsi)
+	popq %rax /* CS: discard */
+	popq %rax /* RFLAGS */
+	movq %rax, _thread_offset_to_rflags(%rsi)
+	popq %rax /* RSP */
+	movq %rax, _thread_offset_to_rsp(%rsi)
+	popq %rax /* SS: discard */
```

Comment:
This is a **third** implementation of context save, with different semantics because you're saving to the thread struct instead of the stack (the original code always put the context on whichever stack it was on).   Again, I don't see the value here vs. having just one block for entry code.  Every new complexity that gets added needs to be fixed in triplicate.

This also has a non-trivial performance cost, as those indirected mov instructions encode using 2-4x as many bytes as a simple push, and the copy of the exception frame at the end is simple waste vs. just leaving it on the stack to begin with.

Reply:
> This is a **third** implementation of context save, with different semantics because you're saving to the thread struct instead of the stack (the original code always put the context on whichever stack it was on). Again, I don't see the value here vs. having just one block for entry code. Every new complexity that gets added needs to be fixed in triplicate.

Remember, we're not `-fno-red-zone` so things are a bit more complicated than they might be on other platforms that can nest on one stack.

There are three context save scenarios, and they're all a bit different:

1. A deliberate __swap() out: this saves **partial** state to the **k_thread**.
2. An unnested IRQ: this saves **complete** state to the **k_thread**.
3. A nested IRQ: this saves **partial** state to the **stack**.

(As you pointed out, I'm saving complete state currently in case 3, but I will fix that.)

The reason case 2 and case 3 differ is because an unnested IRQ may result in thread preemption, and that thread may be resumed from a different CPU and thus we can't rely on any current-CPU state to be correct.  This is obviously not relevant *now*, because SMP doesn't work with `__swap()`, but that's a different discussion.

> This also has a non-trivial performance cost, as those indirected mov instructions encode using 2-4x as many bytes as a simple push

Actually, AMD recommends the MOV over the PUSH: the MOV is easier to parallelize because it has no hidden dependencies (read: RSP).  In general, there's not much of a correlation between the size of an instruction and its execution penalty on these large-core superscalar processors.  If we're writing for a "generic 64-bit target" it's often pointless to engage in these kinds of micro-optimization, because different micro-architectures will do better with different sequences.

> and the copy of the exception frame at the end is simple waste vs. just leaving it on the stack to begin with.

Can't leave it on the stack.  The thread might be preempted, and the interrupt stack is unique to the CPU, not the thread.